src/herder/FetchingEnvelopes.cpp

```diff
@@ -0,0 +1,237 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "herder/FetchingEnvelopes.h"
+#include "crypto/Hex.h"
+#include "crypto/SHA.h"
+#include "herder/Herder.h"
+#include "herder/HerderUtils.h"
+#include "item/ItemFetcher.h"
+#include "main/Application.h"
+#include "overlay/OverlayManager.h"
+#include "overlay/QSetCache.h"
+#include "overlay/TxSetCache.h"
+#include "scp/QuorumSetUtils.h"
+#include "util/Logging.h"
+
+#include <xdrpp/marshal.h>
+
+namespace stellar
+{
+
+using xdr::operator<;
+
+FetchingEnvelopes::FetchingEnvelopes(Application& app)
+    : mApp(app), mNodesInQuorum(mApp)
+{
+}
+
+FetchingEnvelopes::~FetchingEnvelopes()
+{
+}
+
+bool
+FetchingEnvelopes::handleEnvelope(Peer::pointer peer,
+                                  SCPEnvelope const& envelope)
+{
+    if (isDiscarded(envelope))
+    {
+        return false;
+    }
+
+    auto items = itemsToFetch(envelope);
+    if (items.empty())
+    {
+        return true;
+    }
+    else
+    {
+        startFetching(peer, envelope, items);
+        return false;
+    }
+}
+
+std::set<SCPEnvelope>
+FetchingEnvelopes::handleQuorumSet(const SCPQuorumSet& q, bool force)
+{
+    auto itemKey = ItemKey{ItemType::QUORUM_SET, sha256(xdr::xdr_to_opaque(q))};
+    CLOG(TRACE, "Herder") << "Add SCPQSet " << hexAbbrev(itemKey.getHash())
+                          << (force ? ", forced" : "");
+
+    if (!mApp.getItemFetcher().stopFetch(itemKey) && !force)
+    {
+        return std::set<SCPEnvelope>{};
+    }
+
+    if (!isQuorumSetSane(q, false))
+    {
+        discardEnvelopesWithItem(itemKey);
+        return std::set<SCPEnvelope>{};
+    }
+
+    mNodesInQuorum.clear();
+    mApp.getOverlayManager().getQSetCache().add(itemKey.getHash(), q);
```

Comment:
I think this PR would look a lot better if we were not managing the cache directly from herder: this looks to me like something we can delegate to the item fetcher code